src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -89,125 +94,96 @@ internal async ValueTask SetSecurityConfigForConnection(X509Certificate cert, st
-                                connectionEvent);
-                        }
-                        break;
-
-                    // Connection is being closed by the peer
+                        return HandleEventShutdownInitiatedByTransport(ref connectionEvent);
                     case QUIC_CONNECTION_EVENT.SHUTDOWN_INITIATED_BY_PEER:
-                        {
-                            status = HandleEventShutdownInitiatedByPeer(
-                                connectionEvent);
-                        }
-                        break;
-
-                    // Connection has been shutdown
+                        return HandleEventShutdownInitiatedByPeer(ref connectionEvent);
                     case QUIC_CONNECTION_EVENT.SHUTDOWN_COMPLETE:
-                        {
-                            status = HandleEventShutdownComplete(
-                                connectionEvent);
-                        }
-                        break;
-
+                        return HandleEventShutdownComplete(ref connectionEvent);
                     case QUIC_CONNECTION_EVENT.PEER_STREAM_STARTED:
-                        {
-                            status = HandleEventNewStream(
-                                connectionEvent);
-                        }
-                        break;
-
+                        return HandleEventNewStream(ref connectionEvent);
                     case QUIC_CONNECTION_EVENT.STREAMS_AVAILABLE:
-                        {
-                            status = HandleEventStreamsAvailable(
-                                connectionEvent);
-                        }
-                        break;
-
+                        return HandleEventStreamsAvailable(ref connectionEvent);
                     default:
-                        break;
+                        return MsQuicStatusCodes.Success;
                 }
             }
-            catch (Exception)
+            catch (Exception ex)
             {
-                // TODO we may want to either add a debug assert here or return specific error codes
-                // based on the exception caught.
+                if (NetEventSource.Log.IsEnabled())
+                {
+                    NetEventSource.Error(this, $"Exception occurred during connection callback: {ex.Message}");
+                }
+
+                // TODO: trigger an exception on any outstanding async calls.
+
                 return MsQuicStatusCodes.InternalError;
             }
-
-            return status;
         }
 
-        private uint HandleEventConnected(ConnectionEvent connectionEvent)
+        private uint HandleEventConnected(ref ConnectionEvent connectionEvent)
         {
-            SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(MsQuicApi.Api, _ptr, (uint)QUIC_PARAM_LEVEL.CONNECTION, (uint)QUIC_PARAM_CONN.LOCAL_ADDRESS);
-            _localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(inetAddress);
+            if (!_connected)
```

Comment:
Can this ever happen when _connected is true? Should we assert and/or throw?

Reply:
This happens on the listener side -- MsQuic fires the connected event immediately after it fires the new connection event.